File: 0189-rotate-array/0189-rotate-array.java
class Solution {
    public void rotate(int[] nums, int k) {
        int len = nums.length;
        k = k % len; // as k can be greater than the length of the array and effectively number of different results can only upto length of the array. So 'mod'ing k with len. To avoid index out of the bounds exception as well.
        reverseArray(nums, 0, len - 1);
        reverseArray(nums, 0, k - 1);
        reverseArray(nums, k , len - 1);
        
    }
    
    private void reverseArray(int[] nums, int start, int end) {
        while (start < end) {
            int temp = nums[start];
            nums[start] = nums[end];
            nums[end] = temp;
            start++;
            end--;
        }
    }
}

File: 0189-rotate-array/0189-rotate-array.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        // IN-PLACE
        if (k == 0) {
            return;
        }
        int n = nums.size();
        // [1,2,3,4,5], k = 2
        // Step 1: Reverse entire array --> [5,4,3,2,1]
        // Step 2: Reverse the sub-array[0..k - 1] --> [4,5,3,2,1]
        // Step 3: Reverse the sub-array[k.. n - 1] --> [4,5,1,2,3]
        k = (k > n) ? k % n : k;
        reverse(nums.begin(), nums.end());
        reverse(nums.begin(), nums.begin() + k);
        reverse(nums.begin() + k, nums.end());
    }
};